GeosCore/fullchem_mod.F90

```diff
@@ -1130,6 +1132,30 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
 #endif
        ENDIF
 
+#if defined( MODEL_GEOS )
+       ! Check for negative concentrations after first integration and 
+       ! eventually mark integration as errorenous so that it will be
+       ! repeated below. cakelle2, 2023/10/26.
+       IF ( Input_Opt%KppCheckNegatives >= 0 ) THEN
+          IF ( ( Input_Opt%KppCheckNegatives==0 .AND. State_Met%InStratMeso(I,J,L) ) .OR. &
+               ( L > (State_Grid%NZ-Input_Opt%KppCheckNegatives) ) ) THEN
+             IF ( ANY(C < 0.0_dp) ) THEN
+                IERR = -999
+                IF ( State_Diag%Archive_KppError ) THEN
+                   State_Diag%KppError(I,J,L) = State_Diag%KppError(I,J,L) + 1.0 
+                ENDIF
+             ENDIF
+          ENDIF
+       ENDIF
+       IF ( State_Diag%Archive_KppNegatives0 ) THEN
+          S = 0
+          DO N=1,NSPEC
+             IF ( C(N) < 0.0_dp ) S=S+1
+          ENDDO 
+          State_Diag%KppNegatives0(I,J,L) = REAL(S,KIND=f4)
```

Comment:
You can get rid of the loop with an array-based function, which will be more efficient than looping:
```F90 
State_Diag%KppNegatives0(I,J,L) = REAL( COUNT( C < 0.0_dp ), KIND=4 )
```